### PR TITLE
(2) Implement datasources

### PR DESCRIPTION
**Pull request by sylviamoss**
_Should be merged after https://github.com/hashicorp/packer-plugin-sdk/pull/6 [merged ✔️ ]_ 
_Should be merged after https://github.com/hashicorp/packer-plugin-sdk/pull/9 [merged ✔️ ]_
_Should be merged after https://github.com/hashicorp/packer-plugin-sdk/pull/11 [merged ✔️ ]_
_Should be merged after https://github.com/hashicorp/packer-plugin-sdk/pull/15 [merged ✔️ ]_

- [x] implement configure
- [x] Rebase with master
- [x] Move Datasources to https://github.com/hashicorp/packer-plugin-sdk
- [x] Write tests
- [x] Write documentation https://github.com/hashicorp/packer/pull/10489

### CODE REVIEW

File: packer-plugin-sdk/packer/datasource.go
Comment (azr) [inline packer-plugin-sdk/packer/datasource.go:22]: Nice !

File: packer/run_interfaces.go
Comment (azr) [inline packer/run_interfaces.go:33]: Would it be a good idea to make this more explicit ?
```suggestion
	ExecuteDatasources bool
```

File: datasource/amazon/ami/data.go
Comment (azr) [inline datasource/amazon/ami/data.go:19]: A datasource can return nested hcl 'objects' and I think it's really important we work with objects here and I can help with that if you need ! 🙂 
The AMI datasource already should return an object that could look like:

```go
// DataSourceOutput represents the possible fields returned by this datasource.
type DataSourceOutput struct {
  ID           string
  Name         string
  CreationDate string
  Owner        string
  OwnerName    string
  Tags         map[string]string
}
```
One thing we can do here is :

* create a go struct that represents the datasource return type ( just like for a plugin )
* generate the hcl2 spec code for it.
* and use that func here:
```suggestion
	return DataSourceOutput{}.FlatMapstructure().HCL2Spec()
```
* This could also auto-gen some docs for us !!!

Comment (azr) [reply datasource/amazon/ami/data.go:19]: On the same topic, I know that we've talked on zoom about the fact that unknown values and specs don't go through the wire using gob/rpc.

We could add some code to generate a 'mock' value for such a struct so we don't have to fix that, but I think that the time this will take should rather be spent on making this work with unknown values of the type defined by `OutputSpec`. Your call !

File: hcl2template/parser.go
Comment (azr) [inline hcl2template/parser.go:66]: 👍🏼

File: hcl2template/parser.go
Comment (azr) [inline hcl2template/parser.go:266]: Here input variables must be usable in data sources, is that the case ?
Also can I use an input var in a local ?

File: hcl2template/parser.go
Comment (azr) [inline hcl2template/parser.go:308]: I think that we should do a different datasource implementation per datasource type here: 
```suggestion
			placeholderValue := getSpecValue(datasource.OutputSpec())
```

File: hcl2template/types.packer_config.go
Comment (azr) [inline hcl2template/types.packer_config.go:52]: I think this should not be in here but rather returned by the function creating it. The `PackerConfig` represents a loaded hcl config in the hcl sense. Where execution leads to values.

**Review by azr — CHANGES_REQUESTED** [review]
Super nice one ! Great scaffolding, lots of tests and some docs ! 

I also had some thoughts on things ! Everything is in the comments !

🙂 🙂 🙂

File: hcl2template/types.packer_config.go
Comment (sylviamoss) [reply hcl2template/types.packer_config.go:52]: We use the same value more than once, for locals and for builds. We need to keep these save somewhere to avoid executing again and again. I'll try to think of a solution that keeps this behavior.

File: hcl2template/parser.go
Comment (sylviamoss) [reply hcl2template/parser.go:308]: This is because `datasource.OutputSpec()` returns the ObjectSpec which is a map and, as I implemented thinking of a single value, the map will always have 1 element to the type, as I did with the amazon-ami example. 
But making the changes you suggested with multiple values in a struct, this will have to change anyway.

File: packer/run_interfaces.go
Comment (sylviamoss) [reply packer/run_interfaces.go:33]: I think the datasource execution should be the default here, then maybe `SkipDatasourceExecution` 🤔

File: hcl2template/parser.go
Comment (ChiefyG-Admin) [reply hcl2template/parser.go:66]: Well said let me see if I can get any more info on the second pass

File: hcl2template/parser.go
Comment (azr) [reply hcl2template/parser.go:308]: Ah gotcha !

File: hcl2template/types.packer_config.go
Comment (azr) [reply hcl2template/types.packer_config.go:52]: Gotcha ! Yes We need to have some sort of more explicit 'storage' here. 🙂

File: hcl2template/parser.go
Comment (sylviamoss) [reply hcl2template/parser.go:266]: Input variables works on data sources. In `PackerConfig.startDatasource` I decode the datasource block with 
`decodeHCL2Spec(body, cfg.EvalContext(nil), datasource)`

Did you mean to use a data source variable in a local? If it so, the answer is yes.

File: datasource/amazon/ami/data.go
Comment (ChiefyG-Admin) [reply datasource/amazon/ami/data.go:19]: I would very much enjoy learning more about go code. FRom what I've seen it is very useful and a huge asset.

File: packer-plugin-sdk/packer/datasource.go
Comment (SwampDragons) [reply packer-plugin-sdk/packer/datasource.go:22]: yeah this seems great.

**Review by SwampDragons — COMMENTED** [review]
This looks so good! This is there, we just have to make sure we've got docs, an example in the scaffolding repo, and make a "real" version of the amazon-ami datasource.

I think this is super powerful stuff.

File: datasource/amazon/ami/data.go
Comment (azr) [inline datasource/amazon/ami/data.go:28]: Really cool !

File: datasource/amazon/ami/data.go
Comment (sylviamoss) [reply datasource/amazon/ami/data.go:36]: Yes, is just a mock used to test everything. I'm planning to open a new PR with the actual amazon-ami datasource and this PR at the end will contain only the infrastructure stuff.

File: datasource/amazon/ami/data.go
Comment (sylviamoss) [reply datasource/amazon/ami/data.go:28]: I like it too! It's clean.

File: datasource/amazon/ami/data.go
Comment (azr) [inline datasource/amazon/ami/data.go:38]: Is it easy to transform a the DataSource into a cty.Value ? Will we have to add some code generation that will do that ?

Comment (sylviamoss) [reply datasource/amazon/ami/data.go:38]: I have already thought about that, we have to do change [HCL2ValueFromConfigValue(v interface{}) cty.Value](https://github.com/hashicorp/packer/blob/master/hcl2template/shim/values.go#L92) and add the interface case there, and then move this code to the packer-plugin-sdk.

Comment (sylviamoss) [reply datasource/amazon/ami/data.go:38]: I also thought about the code generation but not in the actual implementation 🤔  but we can make use of HCL2ValueFromConfigValue in there too.

Comment (azr) [reply datasource/amazon/ami/data.go:38]: Oh yeah great idea, we might want to put it in the sdk then, and test it way more. 🤔 . ( and test that the output value matches the type from `OutputSpec`. )

Comment (sylviamoss) [reply datasource/amazon/ami/data.go:38]: will add it to the todo list here!

File: cmd/packer-plugin-check/main.go
Comment (nywilken) [inline cmd/packer-plugin-check/main.go:132]: I don't know if the plugin types will grow anytime soon, but I saw this and thought I will probably get the order wrong at some point. What do you think about having `config.GetPlugins()` return a structure, along with an error? 

The structure would have fields for each of the plugin types and the error would only be non-nil when no plugins could be loaded.

Comment (sylviamoss) [reply cmd/packer-plugin-check/main.go:132]: Good idea!! I'll leave the error out in this case because it's literally just a getter of private attributes.

File: hcl2template/types.packer_config.go
Comment (nywilken) [inline hcl2template/types.packer_config.go:274]: Nit: We can use the happy path here and remove the need for the else statement. 

```go
...
		if skipExecution {
			placeholderValue := cty.UnknownVal(hcldec.ImpliedType(datasource.OutputSpec()))
			ds.value = placeholderValue
			cfg.Datasources[ref] = ds
			continue
		} 
		realValue, err := datasource.Execute()
		if err != nil {
			diags = append(diags, &hcl.Diagnostic{
				Summary:  err.Error(),
				Subject:  &cfg.Datasources[ref].block.DefRange,
				Severity: hcl.DiagError,
			})
			continue
		}
	        cfg.Datasources[ref] = realValue
	}

```

**Review by nywilken — COMMENTED** [review]
This looks really good. I left a couple of questions and a nit.

File: config.go
Comment (sylviamoss) [reply config.go:194]: Good catch! Really. With that, I realized that is missing the external datasource plugin discovery. I had the test in my todo list but is good that you point it out before I get that. Thanks!

File: hcl2template/parser.go
Comment (sylviamoss) [reply hcl2template/parser.go:23]: But terraform uses `data` as well. Check it out https://www.terraform.io/docs/configuration/data-sources.html

Comment (nywilken) [reply hcl2template/parser.go:23]: Oops - disregard my comment. I thought the label was datasource sorry for not confirming. Carry on, nothing to see here 😄

File: cmd/packer-plugin-check/main.go
Comment (azr) [reply cmd/packer-plugin-check/main.go:132]: These fields are part of config, I think we can simply remove the `GetPlugins` call and make those public. ( Actually I made them public for #10304 . 🙂

File: cmd/packer-plugin-check/main.go
Comment (azr) [inline cmd/packer-plugin-check/main.go:137]: These fields are part of config, I think we can simply remove the `GetPlugins` call and make those public. ( Actually I made them public for #10304 because I plan on using that config struct everywhere 🙂  )

File: hcl2template/shim/values.go
Comment (azr) [inline hcl2template/shim/values.go]: Yup, panicing sounds harsh but this should only happen in dev mode but I think we should add a comment here for the plugin building people 🙂 .

File: hcl2template/shim/values_test.go
Comment (azr) [inline hcl2template/shim/values_test.go]: Love this tests !

File: hcl2template/shim/values.go
Comment (sylviamoss) [reply hcl2template/shim/values.go]: Add a comment and still panic when this fails?

File: command/hcl2_upgrade.go
Comment (azr) [inline command/hcl2_upgrade.go:15]: I like this name !

File: hcl2template/parser.go
Comment (azr) [inline hcl2template/parser.go:348]: nice !

File: hcl2template/types.datasource.go
Comment (azr) [inline hcl2template/types.datasource.go:71]: I think that if we plan on never adding any other field to a DatasourceRef then, this func is not needed. I had added it to the SourceRef blocks in because I needed to add a body in case the config overloaded the source in a build block, and then if that field was set, then the `SourceRef` would not match in the 'unique source map'.
```suggestion
```

Comment (azr) [reply hcl2template/types.datasource.go:71]: This will be removed in there: https://github.com/hashicorp/packer/pull/10304/files#diff-83936f835c1a77350b68035469f13800efca160475d977415a79ce1f17853670

Comment (azr) [reply hcl2template/types.datasource.go:71]: ( I added an other intermediary type in that could contain the additional block )

File: hcl2template/types.datasource.go
Comment (azr) [inline hcl2template/types.datasource.go:135]: Could you please move this check to `startDatasource`, for #10304 we will check the existence of plugins only in Parse with this PR.  ( I can also do it there if you want ! )

File: hcl2template/types.packer_config.go
Comment (azr) [inline hcl2template/types.packer_config.go:260]: Nice !

File: hcl2template/shim/values.go
Comment (azr) [inline hcl2template/shim/values.go:89]: Nice ! That's currently the only way 👍🏼

**Review by azr — APPROVED** [review]
Love this ! LGTM

File: hcl2template/types.datasource.go
Comment (sylviamoss) [reply hcl2template/types.datasource.go:135]: ✅

File: hcl2template/types.datasource.go
Comment (sylviamoss) [reply hcl2template/types.datasource.go:71]: This ref is being used to check for duplicated data blocks and it makes it very simple by using the ref as a key to look for an existing data source in the map. I see that you still use it to do the same with source blocks. If I remove this I will have to either make the key of the map a string of type+name or make it a list instead and perform a search over a slice to look for duplications.

Adding some sort of `DatasourceUseBlock` will add one unnecessary layer to it because data sources don't have usage specifics like sources. 

Do you have any specific concerns about the DatasourceRef? I think right now it exists only to make it easier to deal with the list of data sources.

File: hcl2template/parser.go
Comment (azr) [inline hcl2template/parser.go:359]:
```suggestion
			if existing, found := cfg.Datasources[*datasource]; found {
```

File: hcl2template/types.datasource.go
Comment (azr) [reply hcl2template/types.datasource.go:71]: no concerns, just that this function is not necessary, a `*DatasourceRef` should not be used as a map accessor, but a `DatasourceRef` can, the map type will know what to do. This function is handy because it returns a `DatasourceRef`.

Comment (azr) [reply hcl2template/types.datasource.go:71]: I made a suggestion here: https://github.com/hashicorp/packer/pull/10440#discussion_r559584486

Comment (sylviamoss) [reply hcl2template/types.datasource.go:71]: I'm looking at your PR now and I can also make it as you did with the RequiredPlugins 🤔

Comment (azr) [reply hcl2template/types.datasource.go:71]: (DatasourceUseBlock will not be necessary here as we never override this like for a source block)

Comment (sylviamoss) [reply hcl2template/types.datasource.go:71]: I just realized that this function wasn't being used anywhere. I think I created it by mistake. But there's still the `(data *Datasource) Ref() DatasourceRef` which is the one being actually used.

I'm not using `*DatasourceRef` as the map accessor 🤔

Comment (azr) [reply hcl2template/types.datasource.go]: ah 🙌🏼  good catch !

File: hcl2template/types.datasource.go
Comment (azr) [inline hcl2template/types.datasource.go:34]: Oh right, I see now ! Nice !

**Comment by SwampDragons** [comment]
looks like this is gonna need a careful rebase. I'm going to leave that to you -- the code in here looks good so once you've rebased, merge.

**Pull request merged**
